t/SocketWithLogging.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <boost/asio/buffers_iterator.hpp>
+#include <boost/system/error_code.hpp>
+#include <iostream>
+#include <utility>
+
+namespace bredis {
+namespace test {
+
+template <typename NextLayer> class SocketWithLogging {
+    NextLayer stream_;
+
+  public:
+    template <typename... Args>
+    SocketWithLogging(Args &&... args) : stream_(std::forward<Args>(args)...) {}
+
+    void cancel() {
+        std::cout << "{cancel}" << std::endl;
+        stream_.cancel();
+    }
+
+    template <typename BufferSequence>
+    void dump(const char *prefix, const BufferSequence &buffers) {
+        using Iterator = boost::asio::buffers_iterator<BufferSequence, char>;
+
+        Iterator it = Iterator::begin(buffers);
+        Iterator end = Iterator::end(buffers);
+        auto size = std::distance(it, end);
+
+        std::cout << "{" << prefix << " " << size << " bytes}[";
+        for (; it != end; it++) {
+            std::cout << *it;
+        }
+        std::cout << "]" << std::endl;
+    }
+
+    template <typename MutableBufferSequence, typename ReadHandler>
+    inline void async_read_some(const MutableBufferSequence &buffers,
+                                ReadHandler handler) {
+        stream_.async_read_some(buffers, handler);
+        dump("async_read_some", buffers);
+    }
+
+    template <typename ConstBufferSequence, typename WriteHandler>
+    inline void async_write_some(const ConstBufferSequence &buffers,
+                                 WriteHandler handler) {
+        dump("async_write_some", buffers);
+        stream_.async_write_some(buffers, handler);
+    }
+
+    template <typename ConstBufferSequence>
+    std::size_t write_some(const ConstBufferSequence &buffers) {
+        dump("write_some", buffers);
+        return stream_.write_some(buffers);
```

Comment:
You've got a bug here. What if `stream_.write_some` doesn't print all the bytes? You will get called again, and then you'll duplicate some of the stuff that was already printed (the tail end of the buffer). You need something like this:
**1**
```
    auto const bytes_transferred = stream_.write_some(buffers);
    dump("write_some", buffers, bytes_transferred);
    return bytes_transferred;
```
OR
**2**
```
    auto const bytes_transferred = stream_.write_some(buffers);
    dump("write_some", beast::prepare_buffers(bytes_transferred, buffers));
    return bytes_transferred;
```